File: nextjs-fdas/src/lib/api/conversation.ts
import { Message, Citation } from '@/types';
import { MessageRequestSchema, ConversationCreateRequestSchema } from '@/validation/schemas';
import { validateRequest } from '../../lib/validation/api-validation';

// API base URL - would be configured based on environment
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api';

/**
 * Conversation API service
 * Handles communication with the backend for conversation operations
 */
class ConversationApiService {
  /**
   * Send a request to the API
   */
  private async request<T>(
    endpoint: string,
    method: string = 'GET',
    data?: any,
    formData?: FormData
  ): Promise<T> {
    // Ensure endpoint starts with / 
    if (!endpoint.startsWith('/')) {
      endpoint = '/' + endpoint;
    }
    
    // Fixed URL construction to prevent duplicated /api
    const finalUrl = API_BASE_URL.endsWith('/api') 
      ? `${API_BASE_URL}${endpoint}`
      : `${API_BASE_URL}/api${endpoint}`;
      
    console.log(`Sending ${method} request to ${finalUrl}`);
    
    // Create request options
    const options: RequestInit = {
      method,
      headers: {
        'Accept': 'application/json'
      }
    };
    
    // Add request body if provided
    if (data) {
      options.headers = {
        ...options.headers,
        'Content-Type': 'application/json'
      };
      options.body = JSON.stringify(data);
    }
    
    // Add form data if provided
    if (formData) {
      // Remove Content-Type header to let the browser set it with the boundary
      if (options.headers && typeof options.headers === 'object') {
        const headers = options.headers as Record<string, string>;
        delete headers['Content-Type'];
      }
      options.body = formData;
    }
    
    try {
      const response = await fetch(finalUrl, options);
      
      // Handle non-OK responses
      if (!response.ok) {
        let errorMessage = `API error: ${response.status} ${response.statusText}`;
        
        // Try to parse error response as JSON
        try {
          const errorData = await response.json();
          if (errorData.detail) {
            if (typeof errorData.detail === 'string') {
              errorMessage = errorData.detail;
            } else if (Array.isArray(errorData.detail)) {
              // Handle Pydantic validation errors
              errorMessage = errorData.detail.map((err: any) => 
                `${err.loc.join('.')}: ${err.msg}`
              ).join(', ');
            } else {
              errorMessage = JSON.stringify(errorData.detail);
            }
          } else {
            errorMessage = JSON.stringify(errorData);
          }
        } catch (e) {
          // If not JSON, try to get text
          try {
            const errorText = await response.text();
            if (errorText) {
              errorMessage = errorText;
            }
          } catch (textError) {
            // Keep the original error message if we can't parse the response
          }
        }
        
        throw new Error(errorMessage);
      }
      
      // Parse the response
      const responseData = await response.json();
      return responseData as T;
    } catch (error) {
      console.error('API request error:', error);
      throw error;
    }
  }
  
  /**
   * Create a new conversation
   */
  async createConversation(title: string = 'New Conversation', documentIds: string[] = []): Promise<{ sessionId?: string; session_id?: string }> {
    try {
      // Validate request data against schema
      const validatedData = validateRequest(ConversationCreateRequestSchema, {
        title,
        documentIds: documentIds,
        userId: 'default-user' // This is handled by the backend, but we'll include it for completeness
      });
      
      const response = await this.request<{ sessionId?: string; session_id?: string }>(
        `/conversation`,
        'POST',
        validatedData
      );
      
      console.log(`Created conversation session: ${response.sessionId || response.session_id}`);
      return response;
    } catch (error) {
      console.error('Error creating conversation:', error);
      throw new Error(`Failed to create conversation: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
  
  /**
   * List conversations
   */
  async listConversations(): Promise<any[]> {
    try {
      const conversations = await this.request<any[]>(
        '/conversation',
        'GET'
      );
      return conversations;
    } catch (error) {
      console.error('Error listing conversations:', error);
      return [];
    }
  }
  
  /**
   * Send a message to the conversation
   */
  async sendMessage(sessionId: string, message: string, documentIds: string[] = []): Promise<Message> {
    if (!sessionId) {
      throw new Error('Session ID is required');
    }

    try {
      // Format the request body to match the backend's expected format
      // The backend expects camelCase due to alias_generator=to_camel
      const requestBody = {
        sessionId: sessionId,
        content: message,
        userId: 'default-user',
        referencedDocuments: documentIds,
        referencedAnalyses: [],
        citationLinks: [],
        citationIds: []
      };
      
      const response = await this.request<Message>(
        `/conversation/${sessionId}/message`,
        'POST',
        requestBody
      );
      
      return response;
    } catch (error) {
      console.error('Error sending message:', error);
      throw new Error(`Failed to send message: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
  
  /**
   * Get messages for a conversation
   */
  async getMessages(sessionId: string, limit: number = 50, offset: number = 0): Promise<Message[]> {
    try {
      const response = await this.request<{ messages: Message[] }>(
        `/conversation/${sessionId}/history?limit=${limit}&offset=${offset}`,
        'GET'
      );
      
      return response.messages;
    } catch (error) {
      console.error('Error getting messages:', error);
      return [];
    }
  }
  
  /**
   * Get document citations
   */
  async getDocumentCitations(documentId: string): Promise<Citation[]> {
    try {
      const citations = await this.request<Citation[]>(
        `/document/${documentId}/citations`,
        'GET'
      );
      
      return citations;
    } catch (error) {
      console.error('Error getting document citations:', error);
      return [];
    }
  }

  /**
   * Add a document to a conversation
   */
  async addDocumentToConversation(conversationId: string, documentId: string): Promise<boolean> {
    try {
      await this.request(
        `/conversation/${conversationId}/document/${documentId}`,
        'POST',
        {} // Empty body, document ID is in the path
      );
      console.log(`Document ${documentId} added to conversation ${conversationId}`);
      return true;
    } catch (error) {
      console.error('Error adding document to conversation:', error);
      return false;
    }
  }

  /**
   * Generate follow-up questions for a conversation
   */
  async generateFollowUpQuestions(conversationId: string, limit: number = 3): Promise<string[]> {
    try {
      const questions = await this.request<string[]>(
        `/conversation/${conversationId}/generate-followups?limit=${limit}`,
        'POST'
      );
      
      console.log(`Generated ${questions.length} follow-up questions for conversation ${conversationId}`);
      return questions;
    } catch (error) {
      console.error('Error generating follow-up questions:', error);
      // Return default questions as fallback
      return [
        "What trends do you see in the financial performance?",
        "How does this compare to industry benchmarks?",
        "What are the key risk factors to consider?"
      ];
    }
  }

  /**
   * Check if a conversation exists
   */
  async checkConversationExists(conversationId: string): Promise<boolean> {
    try {
      await this.request<any>(
        `/conversation/${conversationId}`,
        'GET'
      );
      return true;
    } catch (error) {
      // If we get a 404, the conversation doesn't exist
      if (error instanceof Error && error.message.includes('404')) {
        return false;
      }
      // For other errors, we'll assume it doesn't exist
      console.error('Error checking conversation existence:', error);
      return false;
    }
  }

  /**
   * Get a specific message by ID (with analysis_blocks)
   */
  async getMessage(messageId: string): Promise<Message | null> {
    try {
      const response = await this.request<Message>(
        `/conversation/message/${messageId}`,
        'GET'
      );
      
      return response;
    } catch (error) {
      console.error('Error getting message:', error);
      return null;
    }
  }
}

// Create a singleton instance
export const conversationApi = new ConversationApiService();

File: backend/app/routes/conversation.py
from fastapi import APIRouter, HTTPException, Depends, Query, status
from typing import List, Dict, Any
import logging
from sqlalchemy.ext.asyncio import AsyncSession

from models.message import ConversationCreateRequest, MessageRequest, MessageResponse
from utils.dependencies import get_conversation_service, get_document_service
from models.message import Message, MessageRole, ConversationState
from models.document import Citation
from services.conversation_service import ConversationService
from pdf_processing.api_service import ClaudeService
from pdf_processing.document_service import DocumentService
from repositories.document_repository import DocumentRepository
from repositories.conversation_repository import ConversationRepository
from utils.database import get_db

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/conversation", tags=["conversation"])


# Dependencies
async def get_document_repository(db: AsyncSession = Depends(get_db)):
    return DocumentRepository(db)

async def get_document_service(
    document_repository: DocumentRepository = Depends(get_document_repository)
):
    return DocumentService(document_repository)

async def get_conversation_repository(db: AsyncSession = Depends(get_db)):
    return ConversationRepository(db)

async def get_conversation_service(
    conversation_repository: ConversationRepository = Depends(get_conversation_repository),
    document_repository: DocumentRepository = Depends(get_document_repository)
):
    return ConversationService(
        conversation_repository=conversation_repository,
        document_repository=document_repository
    )

async def get_claude_service():
    return ClaudeService()

async def get_current_user_id():
    # This is a placeholder for authentication
    # In a real application, this would validate the JWT token and return the user ID
    return "default-user"

@router.post("", response_model=ConversationState, response_model_by_alias=True, status_code=status.HTTP_201_CREATED)
async def create_conversation(
    conversation_data: ConversationCreateRequest,
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Create a new conversation with optional document associations.
    """
    try:
        # Override the user_id with the authenticated user
        conversation_data.user_id = user_id
        
        # Create the conversation
        conversation = await conversation_service.create_conversation(
            title=conversation_data.title,
            user_id=conversation_data.user_id,
            document_ids=conversation_data.document_ids or []
        )
        
        # Create and return the conversation state
        state = ConversationState(
            session_id=conversation.id,
            active_documents=conversation_data.document_ids or [],
            last_updated=conversation.updated_at
        )
        
        logger.info(f"Created conversation: {conversation.id} for user: {user_id}")
        return state
    except Exception as e:
        logger.error(f"Error creating conversation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating conversation: {str(e)}"
        )

@router.post("/{session_id}/message", response_model=MessageResponse, response_model_by_alias=True)
async def send_message(
    session_id: str,
    message: MessageRequest,
    conversation_service: ConversationService = Depends(get_conversation_service),
):
    """
    Send a message to a conversation and get an AI response using Claude API.
    The message will be processed and a response generated based on the conversation context.
    If documents are attached to the conversation, they will be used for context.
    Args:
        session_id: The ID of the conversation session
        message: The message content and optional citation IDs
    Returns:
        The AI response message
    """
    if message.session_id != session_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Session ID in path must match session ID in request body"
        )
    try:
        result = await conversation_service.process_user_message(
            conversation_id=session_id,
            content=message.content,
            citation_ids=message.citation_links,
            referenced_documents=message.referenced_documents,
            referenced_analyses=message.referenced_analyses
        )
        if not result.get("success", True):
            logger.warning(f"Error processing message: {result.get('error', 'Unknown error')}")
            # If error, return the assistant message as MessageResponse
            assistant_message = result.get("message")
            if isinstance(assistant_message, dict):
                return MessageResponse(**assistant_message)
            # Convert ORM object to MessageResponse with proper field validation
            return MessageResponse.model_validate(assistant_message, from_attributes=True)
        # Ensure the returned message is a MessageResponse
        message_data = result.get("message")
        if isinstance(message_data, dict):
            return MessageResponse(**message_data)
        # Convert ORM object to MessageResponse with proper field validation
        return MessageResponse.model_validate(message_data, from_attributes=True)
    except ValueError as e:
        logger.error(f"Bad request while processing message: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        logger.exception(f"Error processing message: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"An error occurred while processing the message: {str(e)}"
        )

def is_financial_analysis_request(message: str) -> bool:
    """
    Determine if a message is asking for financial analysis.
    
    Args:
        message: User message text
        
    Returns:
        True if the message appears to be a financial question, False otherwise
    """
    # Simple heuristic to determine if this is a financial question
    financial_keywords = [
        "financial", "revenue", "profit", "margin", "expense", "income",
        "balance sheet", "cash flow", "ratio", "trend", "growth", "forecast",
        "comparison", "analyze", "calculate", "metric", "performance", "quarterly",
        "annual", "fiscal", "q1", "q2", "q3", "q4", "year-over-year", "yoy"
    ]
    
    message_lower = message.lower()
    
    # Check for financial keywords
    if any(keyword in message_lower for keyword in financial_keywords):
        return True
    
    # Check for questions about calculations, comparisons, or visualizations
    calculation_phrases = [
        "calculate", "compare", "show me", "what is the", "how much",
        "visualize", "chart", "graph", "plot", "trend", "over time"
    ]
    
    if any(phrase in message_lower for phrase in calculation_phrases):
        return True
    
    return False

@router.get("/{conversation_id}/history", response_model=List[Message], response_model_by_alias=True)
async def get_conversation_history(
    conversation_id: str,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Get the history of messages for a conversation.
    
    Args:
        conversation_id: ID of the conversation
        limit: Maximum number of messages to return
        offset: Starting offset for pagination
        
    Returns:
        List of messages in the conversation
    """
    # Check if the conversation exists and belongs to the user
    conversation = await conversation_service.get_conversation(conversation_id)
    if not conversation:
        raise HTTPException(status_code=404, detail="Conversation not found")
    
    if conversation.user_id != user_id:
        raise HTTPException(status_code=403, detail="Not authorized to access this conversation")
    
    # Get the conversation messages
    messages = await conversation_service.get_conversation_messages(
        conversation_id=conversation_id,
        limit=limit,
        offset=offset
    )
    
    # Convert messages to API response format
    api_messages = []
    
    for msg in messages:
        # Get citations for this message
        citations = await conversation_service.conversation_repository.get_message_citations(msg.id)
        
        # Convert citations to Citation objects
        citation_objects = []
        for citation in citations:
            # Get the document for this citation
            document = await conversation_service.document_repository.get_document(citation.document_id)
            doc_title = document.filename if document else "Unknown Document"
            
            citation_objects.append(
                Citation(
                    id=citation.id,
                    document_id=citation.document_id,
                    document_title=doc_title,
                    content=citation.content,
                    metadata=citation.metadata or {},
                    type=citation.metadata.get("type", "unknown") if citation.metadata else "unknown"
                )
            )
        
        # Create content blocks if available
        content_blocks = None
        if hasattr(msg, 'content_blocks') and msg.content_blocks:
            content_blocks = msg.content_blocks
        
        # Get analysis blocks for this message
        analysis_blocks = []
        if hasattr(msg, 'analysis_blocks'):
            for block in msg.analysis_blocks:
                analysis_blocks.append({
                    "id": block.id,
                    "block_type": block.block_type,
                    "title": block.title,
                    "content": block.content,
                    "created_at": block.created_at
                })
        
        api_messages.append(
            Message(
                id=msg.id,
                session_id=msg.conversation_id,
                timestamp=msg.created_at,
                role=MessageRole(msg.role),
                content=msg.content,
                referenced_documents=[],  # We don't store this directly in the database
                referenced_analyses=[],   # We don't store this directly in the database
                citation_links=[citation.id for citation in citations],
                citations=citation_objects,
                content_blocks=content_blocks,
                analysis_blocks=analysis_blocks
            )
        )
    
    return api_messages

@router.get("", response_model=List[Dict[str, Any]], response_model_by_alias=True)
async def list_conversations(
    limit: int = Query(10, ge=1, le=50),
    offset: int = Query(0, ge=0),
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    List all conversations for the current user.
    """
    try:
        conversations = await conversation_service.list_conversations(
            user_id=user_id,
            limit=limit,
            offset=offset
        )
        
        # Get the total count
        total_count = await conversation_service.conversation_repository.count_conversations(user_id)
        
        # Format the response
        result = []
        for conversation in conversations:
            # Get the last message for preview
            messages = await conversation_service.get_conversation_messages(
                conversation_id=conversation.id,
                limit=1
            )
            
            last_message = None
            if messages:
                last_message = {
                    "content": messages[0].content[:100] + "..." if len(messages[0].content) > 100 else messages[0].content,
                    "role": messages[0].role,
                    "timestamp": messages[0].created_at.isoformat()
                }
            
            # Get associated documents
            documents = await conversation_service.conversation_repository.get_conversation_documents(conversation.id)
            document_count = len(documents)
            
            result.append({
                "id": conversation.id,
                "title": conversation.title,
                "created_at": conversation.created_at.isoformat(),
                "updated_at": conversation.updated_at.isoformat(),
                "last_message": last_message,
                "document_count": document_count
            })
        
        return result
    except Exception as e:
        logger.error(f"Error listing conversations: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error listing conversations: {str(e)}"
        )

@router.delete("/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_conversation(
    conversation_id: str,
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Delete a conversation and all associated messages.
    """
    try:
        # Check if the conversation exists and belongs to the user
        conversation = await conversation_service.get_conversation(conversation_id)
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Conversation {conversation_id} not found"
            )
        
        if conversation.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to delete this conversation"
            )
        
        # Delete the conversation
        deleted = await conversation_service.conversation_repository.delete_conversation(conversation_id)
        if not deleted:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to delete conversation"
            )
        
        return None
    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except Exception as e:
        logger.error(f"Error deleting conversation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting conversation: {str(e)}"
        )

@router.post("/{conversation_id}/document/{document_id}", status_code=status.HTTP_200_OK)
async def add_document_to_conversation(
    conversation_id: str,
    document_id: str,
    conversation_service: ConversationService = Depends(get_conversation_service),
    document_service: DocumentService = Depends(get_document_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Add a document to a conversation.
    """
    try:
        # Check if the conversation exists and belongs to the user
        conversation = await conversation_service.get_conversation(conversation_id)
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Conversation {conversation_id} not found"
            )
        
        if conversation.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to modify this conversation"
            )
        
        # Check if the document exists and belongs to the user
        document = await document_service.document_repository.get_document(document_id)
        if not document:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Document {document_id} not found"
            )
        
        if document.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to access this document"
            )
        
        # Add the document to the conversation
        success = await conversation_service.conversation_repository.add_document_to_conversation(
            conversation_id=conversation_id,
            document_id=document_id
        )
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to add document to conversation"
            )
        
        return {"message": "Document added to conversation successfully"}
    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except Exception as e:
        logger.error(f"Error adding document to conversation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error adding document to conversation: {str(e)}"
        )

@router.get("/{conversation_id}/document", response_model=List[Dict[str, Any]], response_model_by_alias=True)
async def get_conversation_documents(
    conversation_id: str,
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Get all documents associated with a conversation.
    """
    try:
        # Check if the conversation exists and belongs to the user
        conversation = await conversation_service.get_conversation(conversation_id)
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Conversation {conversation_id} not found"
            )
        
        if conversation.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to access this conversation"
            )
        
        # Get the documents
        documents = await conversation_service.conversation_repository.get_conversation_documents(conversation_id)
        
        # Format the response
        result = []
        for doc in documents:
            result.append({
                "id": doc.id,
                "filename": doc.filename,
                "file_size": doc.file_size,
                "mime_type": doc.mime_type,
                "upload_timestamp": doc.upload_timestamp.isoformat(),
                "document_type": doc.document_type.value if doc.document_type else "other"
            })
        
        return result
    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except Exception as e:
        logger.error(f"Error getting conversation documents: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error getting conversation documents: {str(e)}"
        )

@router.delete("/{conversation_id}/document/{document_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_document_from_conversation(
    conversation_id: str,
    document_id: str,
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
):
    """
    Remove a document from a conversation.
    """
    try:
        # Check if the conversation exists and belongs to the user
        conversation = await conversation_service.get_conversation(conversation_id)
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Conversation {conversation_id} not found"
            )
        
        if conversation.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to modify this conversation"
            )
        
        # Remove the document from the conversation
        success = await conversation_service.conversation_repository.remove_document_from_conversation(
            conversation_id=conversation_id,
            document_id=document_id
        )
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Document not found in conversation"
            )
        
        return None
    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except Exception as e:
        logger.error(f"Error removing document from conversation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error removing document from conversation: {str(e)}"
        )

@router.post("/{conversation_id}/generate-followups", response_model=List[str], response_model_by_alias=True)
async def generate_follow_up_questions(
    conversation_id: str,
    limit: int = Query(3, ge=1, le=5, description="Number of follow-up questions to generate"),
    conversation_service: ConversationService = Depends(get_conversation_service),
    user_id: str = Depends(get_current_user_id)
) -> List[str]:
    """
    Generate contextually relevant follow-up questions based on conversation history.
    
    Args:
        conversation_id: The ID of the conversation
        limit: Number of follow-up questions to generate (1-5, default: 3)
        
    Returns:
        List of follow-up question strings
    """
    try:
        # Check if the conversation exists and belongs to the user
        conversation = await conversation_service.get_conversation(conversation_id)
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Conversation {conversation_id} not found"
            )
        
        if conversation.user_id != user_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to access this conversation"
            )
        
        # Generate follow-up questions
        follow_up_questions = await conversation_service.generate_follow_up_questions(
            conversation_id=conversation_id,
            limit=limit
        )
        
        return follow_up_questions
        
    except HTTPException:
        # Re-raise HTTP exceptions
        raise
    except Exception as e:
        logger.error(f"Error generating follow-up questions: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error generating follow-up questions: {str(e)}"
        )